### PR TITLE
feat: ugc banners

### PR DESCRIPTION
**Pull request by AngeloTadeucci**
- Resolves: #40 
- Resolves: #41 

## Summary by CodeRabbit

- **New Features**
	- Introduced a new `BannerSlot` data management feature, allowing for CRUD operations on banner slots.
	- Enhanced user-generated content (UGC) management with new methods for handling banners, including upload, activation, and loading.

- **Improvements**
	- Improved banner activation logic, ensuring timely activation based on conditions.
	- Streamlined banner upload and validation processes, enhancing usability.

- **Bug Fixes**
	- Enhanced error handling and validation during banner uploads.

- **Documentation**
	- Updated documentation to reflect new features related to banners and UGC management.

- **Refactor**
	- Simplified existing methods for banner management and improved control flows in UGC processing.

### CODE REVIEW

File: Maple2.Server.Web/Controllers/BannerController.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    [HttpGet("{bannerId}/{fileHash}.m2u")]
+    public IResult GetBanner(long bannerId, string fileHash) {
+        string fullPath = Path.Combine(Paths.WEB_DATA_DIR, "banner", bannerId.ToString(), $"{fileHash}.m2u");
+        if (!System.IO.File.Exists(fullPath)) {
+            return Results.BadRequest();
+        }
+
+        FileStream banner = System.IO.File.OpenRead(fullPath);
+        return Results.File(banner, contentType: "application/octet-stream");
+    }
```
[the remark this answers is not in the document]
Comment (AngeloTadeucci) [reply Maple2.Server.Web/Controllers/BannerController.cs]: fixed

File: Maple2.Database/Storage/Game/GameStorage.UgcBannerSlots.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+public partial class GameStorage {
+    public partial class Request {
+        public List<BannerSlot> FindBannerSlotsByBannerId(long bannerId) {
+            return Context.BannerSlots
+                .Where(slot => slot.BannerId == bannerId)
+                .AsEnumerable()
+                .Select(slot => new BannerSlot(slot.Id, slot.ActivateTime, slot.BannerId, slot.Template))
+                .ToList();
+        }
```
[the remark this answers is not in the document]
Comment (AngeloTadeucci) [reply Maple2.Database/Storage/Game/GameStorage.UgcBannerSlots.cs]: fixed

File: Maple2.Server.Core/Packets/UgcPacket.cs [hunk 3]
```diff
@@ -91,14 +104,112 @@ public static ByteWriter UpdateItem(int objectId, Item item, long createPrice) {
...
+                pWriter.WriteUnicodeString();
+                pWriter.WriteByte();
+                pWriter.WriteInt();
+                pWriter.WriteLong();
+                pWriter.WriteLong();
+                pWriter.WriteUnicodeString();
+                pWriter.WriteUnicodeString();
+                pWriter.WriteUnicodeString();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (AngeloTadeucci) [reply Maple2.Server.Core/Packets/UgcPacket.cs:127]: its meant to just document the packet structure

File: Maple2.Model/Metadata/Table/Banner.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+
+namespace Maple2.Model.Metadata;
+
+public record BannerTable(IReadOnlyDictionary<long, BannerTable.Entry> Entries) : Table {
+    public record Entry(
+        long Id,
+        int MapId,
+        List<long> Price) { }
```
Comment (Zintixx) [inline Maple2.Model/Metadata/Table/Banner.cs:9]: Can we make this to an array? We can also remove the curly braces

**Review by Zintixx — CHANGES_REQUESTED** [review]

File: Maple2.Server.Core/Packets/UgcPacket.cs [hunk 5]
```diff
@@ -91,14 +104,112 @@ public static ByteWriter UpdateItem(int objectId, Item item, long createPrice) {
...
+
+        pWriter.WriteInt(banners.Count);
+        foreach (UgcBanner ugcBanner in banners) {
+            pWriter.WriteLong(ugcBanner.Id);
+            BannerSlot? activeSlot = ugcBanner.Slots.FirstOrDefault(x => x.Active);
```
[the remark this answers is not in the document]
Comment (AngeloTadeucci) [reply Maple2.Server.Core/Packets/UgcPacket.cs:132]: I dont think its needed to parameterize this..

UgcBanner is in Maple2.Model so I can't use IUpdatable, should i move UgcBanner to Maple2.Server.Game/Model?

File: Maple2.Server.Core/Packets/UgcPacket.cs [hunk 6]
```diff
@@ -91,14 +104,112 @@ public static ByteWriter UpdateItem(int objectId, Item item, long createPrice) {
...
+            pWriter.WriteLong(bannerSlotDate);
+            pWriter.WriteUnicodeString(bannerSlot.Template?.Author ?? string.Empty);
+            pWriter.WriteBool(true); //  true = reserved, false = awaiting reservation, not sure when false is used
+        }
+    }
+
+    private static void WriteActiveBannerSlot(this ByteWriter pWriter, BannerSlot activeSlot) {
+        pWriter.Write(UgcType.Banner);
+        pWriter.WriteInt(2);
+        pWriter.WriteLong(activeSlot.Template!.AccountId);
```
[the remark this answers is not in the document]
Comment (AngeloTadeucci) [reply Maple2.Server.Core/Packets/UgcPacket.cs:198]: nope slightly different :)

File: Maple2.Server.Core/Packets/UgcPacket.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (AngeloTadeucci) [reply Maple2.Server.Core/Packets/UgcPacket.cs:132]: I can't move to Game/Model, it needs to be used in Maple2.Server.Core and Game isn't a dependency

File: Maple2.Server.Game/PacketHandlers/UgcHandler.cs [hunk 7]
```diff
@@ -170,48 +225,111 @@ private void HandleConfirmation(GameSession session, IByteReader packet) {
...
+                    if (item == null) {
+                        Logger.Fatal("Failed to create UGC Item {ugcUid}", ugcUid);
+                        throw new InvalidOperationException($"Fatal: UGC Item creation: {ugcUid}");
+                    }
+                }
+
+                if (!session.Item.Inventory.CanAdd(item)) {
+                    session.Item.MailItem(item);
+                    return;
+                }
```
Comment (Zintixx) [inline Maple2.Server.Game/PacketHandlers/UgcHandler.cs:260]: Is this return intended? It'll never send the UgcPackets below

Comment (AngeloTadeucci) [reply Maple2.Server.Game/PacketHandlers/UgcHandler.cs:260]: not sure how mail actually works , let me try to replicate and see what happens

File: Maple2.Server.Game/PacketHandlers/UgcHandler.cs [hunk 8]
```diff
@@ -154,12 +154,71 @@
...
         }
 
-        session.StagedUgcItem = item;
+        BannerTable.Entry? bannerMetadata = TableMetadata.BannerTable.Entries.FirstOrDefault(x => x.Id == bannerId);
+        if (bannerMetadata is null) {
+            Logger.Warning("Failed to find banner metadata {BannerId}", bannerId);
+            return;
+        }
+
+        List<BannerSlot> slots = [];
```
[the remark this answers is not in the document]
Comment (AngeloTadeucci) [reply Maple2.Server.Game/PacketHandlers/UgcHandler.cs:176]: Initializing with [] is the new default for c#

[the remark this answers is not in the document]
Comment (AngeloTadeucci) [reply Maple2.Server.Game/PacketHandlers/UgcHandler.cs:176]: It's valid c# and you are incorrect.

[the remark this answers is not in the document]
Comment (mettaursp) [reply Maple2.Server.Game/PacketHandlers/UgcHandler.cs:176]: Shut. Bad bot. Go back to the time out corner

[the remark this answers is not in the document]
Comment (AngeloTadeucci) [reply Maple2.Server.Game/PacketHandlers/UgcHandler.cs:176]: @coderabbitai check this link about the documentation of newer c#, in this case C# 12 in .NET 8
 https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0028?pivots=dotnet-8-0
 
 ```
 // IDE0028 violation.
List<int> list = new List<int>();

// Fixed code (with dotnet_style_prefer_collection_expression = true)
List<int> list = [];
```

File: Maple2.Server.Core/Packets/UgcPacket.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zintixx) [reply Maple2.Server.Core/Packets/UgcPacket.cs:132]: Could we have an implicit converter?

File: Maple2.Server.Game/Model/Field/FieldUgcBanner.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+        BannerSlot? slot = Slots.FirstOrDefault(x => x.ActivateTime.Day == dateTimeOffset.Day && x.ActivateTime.Hour == dateTimeOffset.Hour);
+
+        if (slot is null || slot.Expired || slot.Active) {
+            return;
+        }
+
+        slot.Active = true;
+        field.Broadcast(UgcPacket.ActivateBanner(this));
+    }
```
[the remark this answers is not in the document]
Comment (AngeloTadeucci) [reply Maple2.Server.Game/Model/Field/FieldUgcBanner.cs:26]: that's the same exact code ...

File: Maple2.Server.Game/Model/Field/FieldUgcBanner.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+    private void DeleteOldBannerSlots(DateTimeOffset dateTimeOffset) {
+        foreach (BannerSlot bannerSlot in Slots) {
+            // check if the banner is expired
+            DateTimeOffset expireTimeStamp = dateTimeOffset.Subtract(TimeSpan.FromHours(4));
```
Comment (Zintixx) [inline Maple2.Server.Game/Model/Field/FieldUgcBanner.cs:31]: is this meant to be 4 hours?

Comment (AngeloTadeucci) [reply Maple2.Server.Game/Model/Field/FieldUgcBanner.cs:31]: I think its a value we can tweak 🤔 , dont remember why its 4 hours, just copied from the other emu

Any suggestions?

**Pull request merged**